Add documentation to Colorize class

adlerhsieh opened the pull request.

jhass commented:
That comment doesn't seem to attach anywhere? Or am I wrong, did you check it with `make doc`?

adlerhsieh commented:
I think it is under Colorize::Object class, but I did not check that out with 'make doc' yet. Maybe I'll move to somewhere else if it is not attached to anywhere. Thanks.

adlerhsieh commented:
I followed the [guide](https://github.com/manastech/crystal/blob/master/Contributing.md) to update the docs but could not find what I'v updated including previous ones. I am not sure if my commands are correct?

#### What I did

1. I checked out to `master` branch (up to date) and run `make doc`. No error.
2. Then the `docs/main.cr` [looks like this](https://gist.github.com/adlerhsieh/b544ac2019f34330fc6a), which looks exactly like in the [latest master branch](https://github.com/manastech/crystal/blob/master/docs/main.cr) now.
3. I checked out to `gh-pages` branch and run `rake build && jekyll serve`. No error.
4. On `http://localhost:4000` I found that the new docs did not show up including what I've updated in commit [78693727ff](https://github.com/manastech/crystal/commit/78693727ffa5a4c8ff480e583bf3d36fffe6f3d3) and [bde3711362](https://github.com/manastech/crystal/commit/bde37113624bd05161f2d2c82bed1b390e71c75e). In this case the updated `src/colorize.cr` in this commit did not show up as well.
5. I checked out back to `master` branch, and there is a `_site` directory containing the static files of the docs but they're the same. Not up to date.

#### Commits Confusion

Also, I found that all links in docs are linked to commit [e363b63c05](https://github.com/manastech/crystal/tree/e363b63c05abc63fc11c407c23507ff144900c45). That does not update even after I ran `make doc` and `rake build` commands.

For example, [ECR](http://crystal-lang.org/api/ECR.html) (looks the same on my local machine) links to [the source file](https://github.com/manastech/crystal/blob/e363b63c05abc63fc11c407c23507ff144900c45/src/ecr/ecr.cr) => No docs in the source code. But on my local `master` branch it does contain [the docs I updated](https://github.com/manastech/crystal/blob/master/src/ecr/ecr.cr).

In this case, I tried to update the docs but could not find what I've updated in the source code. Is there any way to make it up to date locally?

adlerhsieh commented:
The documentation should directly attached above the `class` or `module` definition to display. I left a blank line between them so that's why it didn't show up. 

I just added:
```crystal
module Colorize
end
```
in order to attach the documentation right above it and make it visible like this:

![colorize_-_github_com_adlerhsieh_crystal](https://cloud.githubusercontent.com/assets/6851886/10128423/f904522e-65e4-11e5-9ac9-f6af7af10f3e.png)

It took me some time to figure out how `make doc` works and finally it should be visible in the doc now :)

jhass commented on src/colorize.cr:
```diff
@@ -1,3 +1,88 @@
+# The Colorize library allows you to render texts with the given color and text decoration.
```
"With Colorize you can change the fore- and background colors and text decorations when rendering text on terminals supporting ANSI escape codes. It adds the `colorize` method to `Object` and thus all classes as its main interface. Its first argument changes the foreground color:"

jhass commented on src/colorize.cr:
````diff
@@ -1,3 +1,88 @@
+# [1,2,3].colorize(:blue)
+# ```
+#
+# Methods for changing foreground color also include:
````
"There are alternative ways to change the foreground color:"

jhass commented on src/colorize.cr:
````diff
@@ -1,3 +1,88 @@
+# "foo".colorize.green
+# ```
+#
+# Methods for changing background color include:
````
"To change the background color, the following methods are available:"

jhass commented on src/colorize.cr:
````diff
@@ -1,3 +1,88 @@
+# The `colorize` method is included in classes like `String`, `Int32`, `Array`, `Hash`, etc.
+# As default, it accepts a color argument and changes the foreground color:
+# ```crystal
+# "foo".colorize(:green)
````
`require "colorize"`

jhass commented on src/colorize.cr:
````diff
@@ -1,3 +1,88 @@
+# "foo".colorize.on_green
+# ```
+#
+# This library also includes text decoration:
````
"It's also possible to change the text decoration:"

jhass commented on src/colorize.cr:
````diff
@@ -1,3 +1,88 @@
+#
+# Convert to `String` for concatenating.
+# ```crystal
+# "foo".colorize(:red).to_s + " bar"
````
Bad example, `"#{string} interpolation" is preferred over `string + " concatenation"` and calls `to_s` implicitly.

jhass commented on src/colorize.cr:
````diff
@@ -1,3 +1,88 @@
+# #=> "foo" is red
+# ```
+# 
+# You can toggle on/off colors and text decorations.
````
"With the `toggle` method you can temporarily disable adding the escape codes. Settings of the instance are preserved however and can be turned back on later:"

adlerhsieh commented:
Updated.

jhass commented on src/colorize.cr:
````diff
@@ -1,3 +1,91 @@
+# "#{"foo".colorize(:red).to_s} bar"
+# #=> "foo bar"
+# #=> "foo" is red
+# ```
````
I'm not sure we should even mention it, if anything something like "Prefer string interpolation since that calls \`to_s\` for you". Implicitly means it's done for you, "#{foo.to_s} bar" is rather pointless.

adlerhsieh replied:
Agree, I also find that `"#{"foo".colorize(:red)} bar"` works too, which makes the `to_s` method not quite important here. I'll take it out.

adlerhsieh commented:
Took out `to_s` method.
